### DISCUSSION TITLE
Activating oqs-provider in OpenSSL

### DISCUSSION THREAD

**Discussion by octapi345** [open-quantum-safe/oqs-provider#378] [Q&A]
I'm trying to set up oqs-provider on WSL but oqs-provider never becomes activated for OpenSSL. I initially attempted to use the fullbuild.sh and runtests.sh scripts, however the tests always failed.

```
LD_LIBRARY_PATH=/mnt/d/linegraphcode/Provider/oqs-provider/.local/lib
OPENSSL_APP=openssl
OPENSSL_CONF=/mnt/d/linegraphcode/Provider/oqs-provider/scripts/openssl-ca.cnf
OPENSSL_MODULES=/mnt/d/linegraphcode/Provider/oqs-provider/_build/lib
No OQS-OpenSSL111 interop test because of absence of docker
Version information:
OpenSSL 3.1.0 14 Mar 2023 (Library: OpenSSL 3.1.0 14 Mar 2023)
Providers:
  default
    name: OpenSSL Default Provider
    version: 3.1.0
    status: active
    build info: 3.1.0
    gettable provider parameters:
      name: pointer to a UTF8 encoded string (arbitrary size)
      version: pointer to a UTF8 encoded string (arbitrary size)
      buildinfo: pointer to a UTF8 encoded string (arbitrary size)
      status: integer (arbitrary size)
oqsprovider not registered. Exit test.
```
I have also tried the manual cmake commands provided in the Unix install instructions and ran `ctest--rerun-failed --output-on-failure -V` with no errors, but checking openssl with `openssl list -providers` show the same issue as before. The shared library is not being automatically installed in the openssl ./lib/ossl-modules directory. I have tried copying the oqsprovider.so file generated in /_build/lib/ into the ossl-modules directory but this did not fix the issue. Neither did modifying the OPENSSL_MODULES environment variable to point to /_build/lib/. I have modified the openss.conf to have the following. 
```
[provider_sect]
default = default_sect
oqsprovider = oqsprovider_sect
[default_sect]
activate=1
[oqsprovider_sect]
module=oqsprovider.so
activate = 1
```
I attempted using the full file path of oqsprovider.so as the value for module in the config file, but this also did not solve the issue How can I get activate oqs-provider in openssl on WSL?

**Comment by baentsch** [comment]
Thanks for your problem report. I have no personal experience with WSL, so please bear with any possibly unapplicable comments/suggestions below:
- Did the `fullbuild.sh` script execute without errors? Did "oqsprovider.so" get properly created? Did the `liboqs` build execute successfully?
- Is ".so" a proper library suffix in WSL? ".dll" is only for Windows native?
- Are you sure oqsprovider.so is properly placed into the folder pointed to be OPENSSL_MODULES?
- Does `ldd oqsprovider.so` (if that command exists in WSL) indicate no unresolved library references?

**Reply by octapi345** [reply]
As far as i can tell the `oqsprovider.so` file was properly created. I don't know if there are errors within the file but it shows up in the lib folder.` fullbuild.sh` did not show any errors in the terminal. The only potential issue I found is that it listed using a different version of openssl then the one I get from the `openssl version` command (3.0 rather than 3.1). I ended up just doing the setup instructions for Windows native. The issue may have been with the way I set up WSL but I'm not sure. 

Thank you for your help.

**Reply by baentsch** [reply]
> The issue may have been with the way I set up WSL but I'm not sure.

I don't even know how to set up WSL, so I'm afraid I've got pass. But please try to see whether all dependencies resolve correctly (e.g., https://stackoverflow.com/questions/71268653/is-there-a-windows-equivalent-for-the-unix-ldd-command): If they don't, nothing will work.

Then the question whether the OPENSSL_CONF file contains the right activation sequence: Are you certain of that? Please read background in https://github.com/open-quantum-safe/oqs-provider/blob/main/USAGE.md#activation

Another avenue may be to try building [oqs-provider with debug features and analyze the traces](https://github.com/open-quantum-safe/oqs-provider/wiki/Debugging)  -- if any get created.
